clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java

```diff
@@ -36,16 +39,30 @@
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.regex.Pattern;
+import java.util.regex.PatternSyntaxException;
 
 /**
  * Register metrics in JMX as dynamic mbeans based on the metric names
  */
-public class JmxReporter implements MetricsReporter {
+public class JmxReporter implements MetricsReporter, Reconfigurable {
```

Comment:
We might be able to add default implementation for Reconfigurable methods to the MetricsReporter interface, but that would theoretically break binary compatibility for existing plugins.

Reply:
Adding a new method with default implementation to a Java interface doesn't break binary compatibility.  See https://docs.oracle.com/javase/tutorial/java/IandI/defaultmethods.html

> Default methods enable you to add new functionality to the interfaces of your libraries and ensure binary compatibility with code written for older versions of those interfaces.

Reply:
I guess there is a separate question as to whether having a superclass implement a new interface is a binary-compatible change.  Section 13.4.4 of the Java spec seems to imply that it is.  https://docs.oracle.com/javase/specs/jls/se8/html/jls-13.html#jls-13.4.4

> Changing the direct superclass or the set of direct superinterfaces of a class type will not break compatibility with pre-existing binaries, provided that the total set of superclasses or superinterfaces, respectively, of the class type loses no members.

Reply:
According to https://wiki.eclipse.org/Evolving_Java-based_APIs_2#Evolving_API_Interfaces adding a default method is binary incompatible. The reasoning is that if the implementing class already implements a different interface with a default method matching the signature of the added new default methods it would break. In practice this will almost never be the case, but it is theoretically possible. I don't know how strict we are about those things in Kafka.

Reply:
We do consider adding new default methods to interfaces to be a compatible change in Kafka.  I agree that it is theoretically possible for an existing implementation to contain a different  method with the same name, and have an issue that way.  But if we treat every method addition as a compatibility break, it defeats the whole point of default methods in Java.  This would make it very hard to evolve interfaces and would lead to some pretty ugly hacks.

In practice people choose descriptive methods names for public interfaces which makes this is a non-issue.  For example, I think the chances that an existing MetricsReporter implements reconfigurableConfigs, validateReconfiguration, or reconfigure are basically zero.  We also know what most of the widely-deployed MetricsReporters are, so we can check them.

Reply:
@rajinisivaram do I understand correctly that you suggest we leave the Reconfigurable interface optional then?

Reply:
@xvrl: The only reason the interface was optional is that we needed to retain support for Java 7, which didn't have default methods.  Since we don't have to worry about Java 7 any more, let's make the interface part of the type.

Reply:
sorry, I misundertood; I updated the interface to be Reconfigurable and added default methods for backwards compatibility.